File: Sonic4_ModLoader/AMBPatcher/AMB.cs
/*
 * Copyright (c) 2018-2022 Oleg "OSA413" Sokolov
 * Licensed under the MIT License
 * https://github.com/OSA413/Sonic4_ModLoader/blob/master/LICENSE
 */

using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections.Generic;


public class AMB
{
    public enum Version
    {
        PC = 0x20,
        Mobile = 0x28
    }

    private byte[] source;
    public string AmbPath;
    public bool SameEndianness = true;
    public List<BinaryObject> Objects = new List<BinaryObject>();
    public bool hasNames = true;
    public Version version = Version.PC;
    public int Length { get => PredictPointers().name + Objects.Count * (hasNames ? 0x20 : 0) ;}

    public bool IsSourceAMB(int ptr=0)
    {
        return source.Length - ptr >= 0x20
            && source[ptr + 0] == '#'
            && source[ptr + 1] == 'A'
            && source[ptr + 2] == 'M'
            && source[ptr + 3] == 'B';
    }

    public bool IsLittleEndian(byte[] binary = null)
    {
        if (binary == null)
            binary = source;
        bool FileIsLittleEndian = BitConverter.IsLittleEndian;
        if (BitConverter.ToInt32(binary, 4) > 0xFFFF)
            FileIsLittleEndian = !FileIsLittleEndian;
        return FileIsLittleEndian;
    }

    public Version GetVersion(byte[] binary = null, int ptr=0)
    {
        if (binary == null)
            binary = source;

        Version result;

        if (Enum.TryParse<Version>(BitConverter.ToInt32(binary, ptr + 0x4).ToString(), out result))
            return result;
        
        Array.Reverse(binary, ptr + 0x4, 4);
        Enum.TryParse<Version>(BitConverter.ToInt32(binary, ptr + 0x4).ToString(), out result);
        Array.Reverse(binary, ptr + 0x4, 4);

        return result;
    }

    public void SwapEndianness(byte[] binary = null, int ptr=0)
    {
        if (binary == null)
            binary = source;
        var swapHeader = IsLittleEndian(binary) == BitConverter.IsLittleEndian;

        if (!swapHeader)
        {
            Array.Reverse(binary, ptr + 0x4, 4); //version
            Array.Reverse(binary, ptr + 0x10, 4); //file number
            Array.Reverse(binary, ptr + 0x14, 4); //enumeration pointer
            Array.Reverse(binary, ptr + 0x18, 4); //data pointer
            Array.Reverse(binary, ptr + 0x1C, 4); //name pointer
        }

        var objNum = BitConverter.ToInt32(binary, ptr + 0x10);
        var listPointer = BitConverter.ToInt32(binary, ptr + 0x14);

        for (int i = 0; i < objNum; i++)
        {
            Array.Reverse(binary, listPointer + 0x10 * i, 4);
            Array.Reverse(binary, listPointer + 0x10 * i + 4, 4);
        }

        if (swapHeader)
        {
            Array.Reverse(binary, ptr + 0x4, 4);
            Array.Reverse(binary, ptr + 0x10, 4);
            Array.Reverse(binary, ptr + 0x14, 4);
            Array.Reverse(binary, ptr + 0x18, 4);
            Array.Reverse(binary, ptr + 0x1C, 4);
        }
    }

    private (int list, int data, int name) PredictPointers()
    {
        int data = 0x20 + 0x10 * Objects.Count;
        var ptr = data + Objects.Sum(x => x.LengthNice);
        return (0x20, data, hasNames ? ptr : 0);
    }


    private StringBuilder sb = new StringBuilder();
    private string ReadString(byte[] source, int pointer)
    {
        sb.Clear();
        while (pointer < source.Length && source[pointer] != 0x00)
            sb.Append((char)source[pointer++]);

        var str = sb.ToString();
        sb.Clear();
        return str;
    }

    public AMB() {}

    public AMB(string fileName) : this(File.ReadAllBytes(fileName), 0, fileName) {}

    public AMB(byte[] source, int sourcePtr=0, string fileName=null)
    {
        this.source = source;
        AmbPath = fileName;
        if (!IsSourceAMB()) return;
        version = GetVersion();
        
        int shift = 0;
        if (version == Version.Mobile)
            shift = 0x4;

        SameEndianness = BitConverter.IsLittleEndian == IsLittleEndian();

        if (!SameEndianness)
            SwapEndianness();

        var objNum = BitConverter.ToInt32(source, sourcePtr + 0x10);
        var listPtr = BitConverter.ToInt32(source, sourcePtr + 0x14) + sourcePtr;
        //var dataPtr = BitConverter.ToInt32(source, sourcePtr + 0x18 + shift) + sourcePtr; //this may be not dataPtr for mobile
        var namePtr = BitConverter.ToInt32(source, sourcePtr + 0x1C + shift) + sourcePtr;
        if (namePtr == 0) hasNames = false;

        for (int i = 0; i < objNum; i++)
        {
            var objPtr = BitConverter.ToInt32(source, listPtr + (0x10 + shift) * i) + sourcePtr;
            if (objPtr == 0) continue;
            var objLen = BitConverter.ToInt32(source, listPtr + (0x10 + shift) * i + 4 + shift);
            var newObj = new BinaryObject(source, objPtr, objLen);
            newObj.RealName = hasNames ? ReadString(source, namePtr + 0x20 * i) : i.ToString();
            newObj.Name = MakeNameSafe(newObj.RealName);
            newObj.ParentAMB = this;
            if (IsSourceAMB(objPtr))
                newObj.Amb = new AMB(source, objPtr, AmbPath + "\\" + newObj.Name);
            Objects.Add(newObj);
        }

        if (!SameEndianness)
            SwapEndianness();
    }

    public void Save(string filePath = null, bool swapEndianness = false)
    {
        if (filePath == null) filePath = AmbPath;
        if (Path.GetDirectoryName(filePath) != "")
            Directory.CreateDirectory(Path.GetDirectoryName(filePath));
        File.WriteAllBytes(filePath, Write(swapEndianness));
    }

    public byte[] Write(bool swapEndianness = false)
    {
        var result = new byte[Length];
        var pointers = PredictPointers();

        Array.Copy(Encoding.ASCII.GetBytes("#AMB"), 0, result, 0, 4);
        Array.Copy(BitConverter.GetBytes(Enum.IsDefined(typeof(Version), version) ? (int)version : 0x20), 0, result, 0x04, 4); //AMB file version
        Array.Copy(BitConverter.GetBytes(Objects.Count), 0, result, 0x10, 4);
        Array.Copy(BitConverter.GetBytes(pointers.list), 0, result, 0x14, 4);
        Array.Copy(BitConverter.GetBytes(pointers.data), 0, result, 0x18, 4);
        Array.Copy(BitConverter.GetBytes(pointers.name), 0, result, 0x1C, 4);

        foreach (var o in Objects)
        {
            Array.Copy(BitConverter.GetBytes(pointers.data), 0, result, pointers.list, 4);
            Array.Copy(BitConverter.GetBytes(o.LengthNice), 0, result, pointers.list + 4, 4);

            var oWrite = o.Write();
            Array.Copy(oWrite, 0, result, pointers.data, o.Length);
            if (hasNames)
            {
                Array.Copy(Encoding.ASCII.GetBytes(o.RealName), 0, result, pointers.name, o.RealName.Length);
                pointers.name += 0x20;
            }

            pointers.list += 0x10;
            pointers.data += o.LengthNice;
        }

        if (swapEndianness == SameEndianness)
            SwapEndianness(result);

        return result;
    }

    public string GetRelativeName(string MainFileName, string objectName)
    {
        //Turning "C:\1\2\3" into {"C:","1","2","3"}
        var modPathParts = objectName.Replace('/', '\\').Split('\\');
        var origFileName = Path.GetFileName(MainFileName);

        //Trying to find where the original file name starts in the mod file name.
        int index = Array.IndexOf(modPathParts, origFileName);
                    
        string InternalName;
        //If it's inside, return the part after original file ends
        if (index != -1)
            InternalName = String.Join("\\", modPathParts.Skip(index + 1).ToArray());
        //Else use file name
        else
            InternalName = modPathParts.Last();

        //This may occur when main file and added file have the same name
        if (InternalName == "")
            InternalName = modPathParts.Last();

        return InternalName;
    }

    public void Add(string filePath, string newName = null)
    {
        var target = FindObject(newName?.Replace('/', '\\') ?? GetRelativeName(AmbPath, filePath.Replace("_extracted", "")));

        var newObj = new BinaryObject(filePath);
        newObj.Name = newObj.RealName = target.name;

        if (target.index == -1)
            target.amb.Objects.Add(newObj);
        else
            target.amb.Replace(newObj, target.index);
    }

    public (AMB amb, int index, string name) FindObject(string objectName)
    {
        var InternalName = objectName;

        var InternalIndex = Objects.FindIndex(x => x.Name == InternalName);
        if (InternalIndex != -1)
            return (this, InternalIndex, InternalName);

        var ParentIndex = Objects.FindIndex(x => x.Name == String.Join("\\", InternalName.Split('\\').Take(x.Name.Count(c => c == '\\') + 1)));

        if (ParentIndex != -1)
            return Objects[ParentIndex].Amb.FindObject(objectName.Substring(Objects[ParentIndex].Name.Length + 1));

        return (this, -1, InternalName);
    }

    public void Replace(BinaryObject bo, int targetIndex)
    {
        bo.RealName = Objects[targetIndex].RealName;
        Objects[targetIndex] = bo;
    }

    public void ExtractAll(string output = null) => Extract(output, true);

    public void Extract(string output = null, bool extractAll = false)
    {
        if (!IsSourceAMB()) return;
        if (output == null) output = AmbPath + "_extracted";
        Directory.CreateDirectory(output);

        foreach (var o in Objects)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(Path.Combine(output, o.Name)));
            if (extractAll && o.isAMB)
                o.Amb.Extract(Path.Combine(output, o.Name), true);
            else
                o.Save(Path.Combine(output, o.Name));
        }
    }

    public static string MakeNameSafe(string rawName)
    {
        //removing ".\" in the names (Windows can't create "." folders)
        //sometimes they can have several ".\" in the names
        //Turns out there's a double dot directory in file names
        //And double backslash in file names
        int safeIndex = 0;
        while (rawName[safeIndex] == '.' || rawName[safeIndex] == '\\' || rawName[safeIndex] == '/')
            safeIndex++;

        return rawName.Substring(safeIndex);
    }

    public void Remove(int index) => Objects.RemoveAt(index);
    public void Remove(string objectName)
    {
        var target = FindObject(objectName);
        target.amb.Remove(target.index);
    }
}

public class BinaryObject
{
    public string Name;
    public string RealName;
    public bool isAMB { get => Amb != null; }
    public AMB Amb;
    public AMB ParentAMB;

    public byte[] Source {get; private set;}
    //TODO: use uint
    public int Pointer {get; private set;}
    int length;
    public int Length {get => isAMB ? Amb.Length : length;}
    //This makes length of the mod file to be % 16 = 0 with no empty space at the end
    public int LengthNice {get => Length + (16 - Length % 16) % 16;}

    public BinaryObject(byte[] source, int pointer, int length)
    {
        Source = source;
        Pointer = pointer;
        this.length = length;
    }

    public BinaryObject(byte[] source)
    {
        Source = source;
        this.length = Source.Length;
    }

    public BinaryObject(string filePath)
    {
        Source = File.ReadAllBytes(filePath);
        this.length = Source.Length;
    }

    public byte[] Write()
    {
        if (isAMB)
            return Amb.Write();
        return Source.Skip(Pointer).Take(Length).ToArray();
    }

    public void Save(string path) => File.WriteAllBytes(path, Write());
}

File: src/amb-rs/src/amb.rs
use std::{io::Read, path::Path};
use crate::{binary_reader::{self, Endianness}, binary_writer};

pub enum Version {
    PC = 0x20,
    Mobile = 0x28,
}

pub struct Amb {
    pub amb_path: String,
    pub endianness: Option<Endianness>,
    pub objects: Vec<BinaryObject>,
    pub has_names: bool,
    pub version: Version,
}

struct AmbPointersPrediction {
    pub list: usize,
    pub data: usize,
    pub name: usize,
}

impl Amb {
    pub fn length(&self) -> usize {
        self.predict_pointers().name + self.objects.len() * match self.has_names {
            true => 0x20,
            false => 0
        }
    }

    fn predict_pointers(&self) -> AmbPointersPrediction {
        let data = 0x20 + 0x10 * self.objects.len();
        let ptr = data + self.objects.iter().map(|object| object.length_nice()).sum::<usize>();
        AmbPointersPrediction {
            list: 0x20, 
            data: data,
            name: match self.has_names { 
                true => ptr,
                false => 0
            }
        }
    }

    pub fn is_source_amb(source: &Vec<u8>, ptr: Option<usize>) -> bool {
        let ptr = ptr.unwrap_or(0);
        source.len() - ptr >= 0x20
            && source[ptr + 0] == '#' as u8
            && source[ptr + 1] == 'A' as u8
            && source[ptr + 2] == 'M' as u8
            && source[ptr + 3] == 'B' as u8
    }

    pub fn get_version(source: &Vec<u8>, ptr: Option<usize>) -> (Version, Option<Endianness>) {
        let ptr = ptr.unwrap_or(0);
        match binary_reader::read_u32(source, ptr + 0x4, &None).unwrap() {
            0x20 => (Version::PC, Some(Endianness::Little)),
            0x28 => (Version::Mobile, Some(Endianness::Little)),
            value => {
                match value.swap_bytes() {
                    0x20 => (Version::PC, Some(Endianness::Big)),
                    0x28 => (Version::Mobile, Some(Endianness::Big)),
                    _ => {
                        println!("Could not detect version of AMB file");
                        (Version::PC, None)
                    },
                }
            }
        }
    }

    pub fn new_empty() -> Self {
        Self {
            amb_path: String::new(),
            endianness: Some(Endianness::Little),
            objects: Vec::new(),
            has_names: true,
            version: Version::PC,
        }
    }

    pub fn new_from_file_name(file_path: &String) -> Result<Self, std::io::Error> {
        Ok(Self::new_from_src_ptr_name(&std::fs::read(&file_path)?, Some(0), file_path.to_string()))
    }

    pub fn new_from_src_ptr_name(
        source: &Vec<u8>,
        ptr: Option<usize>,
        name: String
    ) -> Self {
        if !Amb::is_source_amb(source, ptr) {
            panic!("Provided source is not an AMB file");
        }
        let ptr = ptr.unwrap_or(0);
        let (version, endianness) = Amb::get_version(source, Some(ptr));
        let shift: usize = match version {
            Version::Mobile => 0x4,
            _ => 0,
        };

        let object_number = binary_reader::read_u32(source, ptr + 0x10, &endianness).expect("Another bad thing happened that you didn't account for #10");
        let list_pointer = binary_reader::read_u32(source, ptr + 0x14, &endianness).expect("Another bad thing happened that you didn't account for #11") + ptr as u32;
        //var dataPtr = binary_reader::read_u32(source, sourcePtr + 0x18 + shift) + sourcePtr; //this may be not dataPtr for mobile
        let names_pointer = binary_reader::read_u32(source, ptr + 0x1C + shift, &endianness).expect("Another bad thing happened that you didn't account for #12") + ptr as u32;
        let has_names = names_pointer != 0;

        let mut objects = Vec::<BinaryObject>::new();
        let mut i: usize = 0;
        while i < object_number as usize {
            let object_pointer = binary_reader::read_u32(source, list_pointer as usize + (0x10 + shift) * i, &endianness).expect("Who's bad?") + ptr as u32;
            if object_pointer == 0 {
                i += 1;
                continue;
            }
            let object_length = binary_reader::read_u32(source, list_pointer as usize + (0x10 + shift) * i + 4 + shift, &endianness).expect("Who's bad? (2)");
            let mut new_object = BinaryObject::new_from_src_ptr_len(source, object_pointer as usize, object_length as usize);
            new_object.real_name = match has_names {
                true => binary_reader::read_string32(source, names_pointer as usize + 0x20 * i),
                false =>  i.to_string(),
            };
            new_object.name = Amb::make_name_safe(&new_object.real_name);
            new_object.flag1 = binary_reader::read_u32(source, list_pointer as usize + (0x10 + shift) * i + 8 + shift, &endianness).expect("Who's bad? (3)");
            new_object.flag2 = binary_reader::read_u32(source, list_pointer as usize + (0x10 + shift) * i + 12 + shift, &endianness).expect("Who's bad? (4)");
            if Amb::is_source_amb(source, Some(object_pointer as usize)) {
                new_object.amb = Some(Amb::new_from_src_ptr_name(source, Some(object_pointer as usize), name.clone() + "\\" + &new_object.name));
            }
            objects.push(new_object);
            i += 1;
        }

        Amb {
            amb_path: name,
            endianness,
            objects,
            has_names,
            version,
        }
    }

    pub fn new_from_binary_object(bo: &BinaryObject) -> Self {
        Amb::new_from_src_ptr_name(&bo.data, Some(bo.pointer), bo.name.clone())
    }

    pub fn write(&self) -> Vec<u8> {
        let mut result = Vec::<u8>::with_capacity(self.length());
        let mut pointers = self.predict_pointers();

        "#AMB".as_bytes().read_exact(&mut result[0x0..0x4]);
        binary_writer::write_u32(&mut result, 0x4, match self.version {
            Version::PC => 0x20,
            Version::Mobile => 0x28,
        }, &self.endianness);
        binary_writer::write_u32(&mut result, 0x10, self.objects.iter().count() as u32, &self.endianness);
        binary_writer::write_u32(&mut result, 0x14, pointers.list as u32, &self.endianness);
        binary_writer::write_u32(&mut result, 0x18, pointers.data as u32, &self.endianness);
        binary_writer::write_u32(&mut result, 0x1C, pointers.name as u32, &self.endianness);

        for o in self.objects.iter() {
            binary_writer::write_u32(&mut result, pointers.list, pointers.data as u32, &self.endianness);
            binary_writer::write_u32(&mut result, pointers.list + 4, o.length_nice() as u32, &self.endianness);
            binary_writer::write_u32(&mut result, pointers.list + 8, o.flag1 as u32, &self.endianness);
            binary_writer::write_u32(&mut result, pointers.list + 12, o.flag2 as u32, &self.endianness);

            let object_data = &o.data;
            result[pointers.data..pointers.data + o.length()].copy_from_slice(&object_data);
            if self.has_names {
                o.real_name.as_bytes().read_exact(&mut result[pointers.name..pointers.name + o.real_name.len()]);
                pointers.name += 0x20;
            }

            pointers.list += 0x10;
            pointers.data += o.length_nice();
        }

        result
    }

    pub fn get_relative_name(&self, main_file_name: String, object_name: String) -> String {
        //Turning "C:\1\2\3" into {"C:","1","2","3"}
        let mod_path_parts = object_name.replace('/', "\\");
        let mod_path_parts = mod_path_parts.split('\\').collect::<Vec<_>>();
        let orig_file_name = Path::new(&main_file_name).file_name().expect("That one bad thing happened that you thought would not happen #1");

        //Trying to find where the original file name starts in the mod file name.
        let index = mod_path_parts.iter().position(|&r| r == orig_file_name);

        let internal_name = match index {
            //If it's inside, return the part after original file ends
            Some(index) => mod_path_parts.iter().skip(index + 1).map(|x| x.to_owned()).collect::<Vec<_>>().join("\\"),
            //Else use file name
            None => mod_path_parts.last().expect("That one bad thing happened that you thought would not happen #2").to_string(),
        };

        //This may occur when main file and added file have the same name
        if internal_name == "" {
            return mod_path_parts.last().expect("That one bad thing happened that you thought would not happen #3").to_string();
        }

        return internal_name;
    }

    pub fn add_binary_object(&mut self, mut new_obj: BinaryObject, internal_name: String) {
        match self.objects.iter().position(|x| x.name == internal_name) {
            Some(internal_index) => {
                let existing_object = &self.objects[internal_index];
                new_obj.name = existing_object.name.clone();
                new_obj.real_name = existing_object.real_name.clone();
                
                self.replace(new_obj, internal_index);
                return;
            },
            None => (),
        }

        match self.objects.iter().position(|x| x.name == internal_name.split('\\').take(x.name.chars().filter(|&c| c == '\\').count() + 1).collect::<Vec<_>>().join("\\")) {
            Some(parent_index) => {
                let parent_object = &self.objects[parent_index];
                // Why do we consider that this object is an AMB?
                let mut parent_amb = Amb::new_from_binary_object(parent_object);
                // ????????
                return parent_amb.add_binary_object(new_obj, internal_name);
            }
            None => todo!(),
        }
    }

    pub fn add(&mut self, file_path: String, new_name: Option<String>) {
        let mut new_obj = BinaryObject::new_from_file_path(file_path.clone()).expect("This happend #6");

        let internal_name = new_name.unwrap_or(self.get_relative_name(self.amb_path.clone(), file_path.replace("_extracted", ""))).replace('/', "\\");

        match self.objects.iter().position(|x| x.name == internal_name) {
            Some(internal_index) => {
                let existing_object = &self.objects[internal_index];
                new_obj.name = existing_object.name.clone();
                new_obj.real_name = existing_object.real_name.clone();
                
                self.objects[internal_index] = new_obj;
                return;
            },
            None => (),
        }

        match self.objects.iter().position(|x| x.name == internal_name.split('\\').take(x.name.chars().filter(|&c| c == '\\').count() + 1).collect::<Vec<_>>().join("\\")) {
            Some(parent_index) => {
                let parent_object = &self.objects[parent_index];
                // Why do we consider that this object is an AMB?
                let mut parent_amb = Amb::new_from_binary_object(parent_object);
                return parent_amb.add_binary_object(new_obj, internal_name);
            }
            None => todo!(),
        }
    }

    pub fn replace(&mut self, mut bo: BinaryObject, target_index: usize) {
        let prev = &self.objects[target_index];
        bo.real_name = prev.real_name.clone();
        bo.flag1 = prev.flag1;
        bo.flag2 = prev.flag2;
        self.objects.remove(target_index);
        self.objects.insert(target_index, bo.try_into().expect("That one bad thing happened that you thought would not happen #5"));
    }

    pub fn make_name_safe(raw_name: &String) -> String {
        //removing ".\" in the names (Windows can't create "." folders)
        //sometimes they can have several ".\" in the names
        //Turns out there's a double dot directory in file names
        //And double backslash in file names
        let mut safe_index = 0;
        // Why mut?
        let mut chars = raw_name.chars();
        while let Some(ch) = chars.nth(safe_index) {
            if ch == '.' || ch == '\\' || ch == '/' {
                safe_index += 1;
            } else {
                break;
            }
        }

        return raw_name.chars().skip(safe_index).collect();
    }

    // Make recursive as `add`?
    pub fn remove(&mut self, object_name: String) {
        let target = self.objects.iter().position(|x| x.name == object_name);
        if let Some(target) = target {
            self.objects.remove(target);
        }
    }
}

pub struct BinaryObject {
    pub name: String,
    pub real_name: String,
    pub amb: Option<Amb>,

    pub flag1: u32,
    pub flag2: u32,

    pub pointer: usize, //This is used just for the output and debugging
    pub data: Vec<u8>,
}

impl BinaryObject {
    pub fn length(&self) -> usize {
        return self.data.len();
    }

    pub fn length_nice(&self) -> usize {
        self.length() + (16 - self.length() % 16) % 16
    }

    pub fn new_from_src_ptr_len(
        source: &Vec<u8>,
        pointer: usize,
        length: usize
    ) -> Self {
        BinaryObject {
            data: source.iter().skip(pointer).take(length).map(|x| x.to_owned()).collect(),
            flag1: 0,
            flag2: 0,
            pointer,
            amb: None,
            name: String::new(),
            real_name: String::new(),
        }
    }

    pub fn new_from_file_path(
        file_path: String
    ) -> Result<Self, std::io::Error> {
        let file_content = std::fs::read(file_path)?;
        Ok(BinaryObject {
            data: file_content,
            flag1: 0,
            flag2: 0,
            pointer: 0,
            amb: None,
            name: String::new(),
            real_name: String::new(),
        })
    }
}